test/evaluate/ISO-Fortran-binding.cc

```diff
@@ -210,14 +210,459 @@ int main() {
+  int numErr{0};
+  int expectedRetCode{CFI_SUCCESS};
+  if (rank < 0 || rank > CFI_MAX_RANK) {
+    ++numErr;
+    expectedRetCode = CFI_INVALID_RANK;
+  }
+  if (type < 0 || type > CFI_type_struct) {
+    ++numErr;
+    expectedRetCode = CFI_INVALID_TYPE;
+  }
+  if (base_addr != nullptr && attribute == CFI_attribute_allocatable) {
+    // This is less restrictive than 18.5.5.3 arg req for which pointers arg
+    // shall be unassociated. However, this match ALLOCATE behavior
+    // (9.7.3/9.7.4)
+    ++numErr;
+    expectedRetCode = CFI_ERROR_BASE_ADDR_NOT_NULL;
+  }
+  if (attribute != CFI_attribute_pointer &&
+      attribute != CFI_attribute_allocatable) {
+    ++numErr;
+    expectedRetCode = CFI_INVALID_ATTRIBUTE;
+  }
+  if (rank > 0 && (lower_bounds == nullptr || upper_bounds == nullptr)) {
+    ++numErr;
+    expectedRetCode = CFI_INVALID_EXTENT;
+  }
+
+  // Memory allocation failures are unpredictable in this test.
+  if (numErr == 0 && retCode != CFI_SUCCESS) {
+    MATCH(true, retCode == CFI_ERROR_MEM_ALLOCATION);
+  } else if (numErr > 1) {
+    MATCH(true, retCode != CFI_SUCCESS);
+  } else {
+    MATCH(expectedRetCode, retCode);
+  }
+  // clean-up
+  if (retCode == CFI_SUCCESS) {
+    CFI_deallocate(dv);
+  }
+}
 
-  // TODO: test CFI_address
-  // TODO: test CFI_allocate
+static void run_CFI_allocate_tests() {
+  // 18.5.5.3
+  // create test world
+  CFI_CDESC_T(CFI_MAX_RANK) dv_storage;
+  CFI_cdesc_t *dv{reinterpret_cast<CFI_cdesc_t *>(&dv_storage)};
+  static char base;
+  void *dummyAddr = reinterpret_cast<void *>(&base);
+  CFI_attribute_t attrCases[]{
+      CFI_attribute_pointer, CFI_attribute_allocatable, CFI_attribute_other};
+  CFI_type_t typeCases[]{CFI_type_int, CFI_type_struct, CFI_type_double,
+      CFI_type_char, CFI_type_other, CFI_type_struct + 1};
+  void *baseAddrCases[]{dummyAddr, static_cast<void *>(nullptr)};
+  CFI_rank_t rankCases[]{0, 1, CFI_MAX_RANK, CFI_MAX_RANK + 1};
+  std::size_t lenCases[]{0, 42};
+  CFI_index_t lb1[CFI_MAX_RANK];
+  CFI_index_t ub1[CFI_MAX_RANK];
+  for (int i{0}; i < CFI_MAX_RANK; ++i) {
+    lb1[i] = -1;
+    ub1[i] = 0;
+  }
+
+  check_CFI_establish(
+      dv, nullptr, CFI_attribute_other, CFI_type_int, 0, 0, nullptr);
+  for (CFI_type_t type : typeCases) {
+    std::size_t ty_len = ByteSize(type, 12);
+    for (CFI_attribute_t attribute : attrCases) {
+      for (void *base_addr : baseAddrCases) {
+        for (CFI_rank_t rank : rankCases) {
+          for (size_t elem_len : lenCases) {
+            dv->base_addr = base_addr;
+            dv->rank = rank;
+            dv->attribute = attribute;
+            dv->type = type;
+            dv->elem_len = ty_len;
+            check_CFI_allocate(dv, lb1, ub1, elem_len);
+          }
+        }
+      }
+    }
+  }
+}
+
+static void run_CFI_section_tests() {
+  // simple tests
+  bool testPreConditions{true};
+  constexpr CFI_index_t m = 5, n = 6, o = 7;
+  constexpr CFI_rank_t rank = 3;
```

Comment:
More curly brace opportunities

Reply:
Thanks, I went through the whole file again to fixed all similar cases.